File: Data Structures & Algorithms/merge-intervals/submission-2.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        max_end = [-1] * (max(i[0] for i in intervals) + 1)

        for start, end in intervals:
            max_end[start] = max(max_end[start], end)

        res = []
        curr_start = None
        curr_end = None
        for i in range(len(max_end)):
            if max_end[i] != -1:
                if curr_start is None:
                    curr_start = i
                    curr_end = max_end[i]
                else:
                    # already in an interval
                    curr_end = max(curr_end, max_end[i])

            if i == curr_end:
                print(curr_start, curr_end)
                if curr_start != -1 and curr_end != -1:
                    res.append([curr_start, curr_end])
                    curr_start = None
                    curr_end = None

        if curr_start is not None:
            res.append([curr_start, curr_end])

        return res

File: Data Structures & Algorithms/meeting-schedule/submission-2.cpp
/**
 * Definition of Interval:
 * class Interval {
 * public:
 *     int start, end;
 *     Interval(int start, int end) {
 *         this->start = start;
 *         this->end = end;
 *     }
 * }
 */

class Solution {
public:
    bool canAttendMeetings(vector<Interval>& intervals) {
        if (intervals.size() <= 1) return true;

        std::sort(intervals.begin(), intervals.end(), [](const auto& a, const auto& b){
            return a.start < b.start;
        });
        for (size_t i = 1; i < intervals.size(); ++i) {
            if (intervals[i-1].end > intervals[i].start) return false;
        }
        return true;
    }
};
